content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+    "goStatementLoc": {
+      "pc": 4387029,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 240,
+      "function": {
+        "name": "runtime.init.4",
+        "value": 4386976,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "startLoc": {
+      "pc": 4387056,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 243,
+      "function": {
+        "name": "runtime.forcegchelper",
+        "value": 4387056,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "threadID": 0
+  },
+  ...
+```
+
+Every object in the JSON represents a single goroutine. By checking [“goroutines” command help](https://github.com/derekparker/delve/blob/master/Documentation/cli/README.md#goroutines) we can figure out what data Delve knows. We're interested in `userCurrentLoc` field, which is "topmost stackframe in user code".
+
+Of all goroutines in the JSON let's list unique function names with the exact line numbers where function got paused:
+
+```
+= jq -c '.result[] | [.userCurrentLoc.function.name, .userCurrentLoc.line]' server-test-1_dlv-rpc-list_goroutines.json | sort | uniq -c
+
+   1 ["internal/poll.runtime_pollWait",173]
+1000 ["main.(*Metrics).CountS",113]
+   1 ["main.(*Metrics).SetM",145]
+   1 ["main.(*Metrics).startOutChannelConsumer",239]
+   1 ["main.run",56]
+   1 ["os/signal.signal_recv",139]
+   6 ["runtime.gopark",303]
+```
+
+The majority of goroutines (1000) have stacked in `main.(*Metrics).CountS:113`. Now, this is the perfect time to look at the source code.
+
+In the `main` package, find `Metrics` struct and look at its `CountS` method (see `example/server/metrics.go`):
+
+```go
+// CountS increments counter per second.
+func (m *Metrics) CountS(key string) {
+    // ···
+
+    m.inChannel <- NewCountMetric(key, 1, second)
+}
+```
+
+Our server has stacked on sending to the `inChannel` channel. Let’s find out who is supposed to read from this channel. After inspecting the code, we should find the following function (example/server/metrics.go):
+
+```
+// starts a consumer for inChannel
+func (m *Metrics) startInChannelConsumer() {
+    for inMetrics := range m.inChannel {
+   	    // ···
+    }
+}
+```
+
+The function reads values out of the channel and does something with them, one by one. In what possible situations, could the sending to this channel being blocked?
+
+When working with channels, there are only four possible "oopsies", according to Dave Cheney's [Channel Axioms](https://dave.cheney.net/2014/03/19/channel-axioms):
+
+- send to a nil channel block forever
+- receive from a nil channel block forever
+- send to a closed channel panics
+- receive from a closed channel returns the zero value immediately.
+
+"Send to a nil channel block forever" – at first sight, this sounds like something possible. But, after double-checking with the code, `inChannel` is initialised in the `Metrics` constructor. So it can't be nil.
```

Comment:
* Add a link to the constructor? https://github.com/narqo/postmortem-debug-go/blob/2c42ca73ebd500fe8da1c6ac8ecaf4af143aca78/example/server/metrics.go#L75
 * "at first sight" and "sounds like" mixes two different metaphors, how about:
```suggestion
"Send to a nil channel block forever" – at first sight, this seems like something possible. But, after double-checking with the code, `inChannel` is initialised in the `Metrics` constructor. So it can't be nil.
```